framework/decode/vulkan_virtual_swapchain.cpp

```diff
@@ -0,0 +1,789 @@
+        offsets[1].y     = static_cast<int32_t>(swapchain_info->height);
+        VkImageBlit blit = {
+            subresource,
+            { offsets[0], offsets[1] },
+            subresource,
+            { offsets[0], offsets[1] },
+        };
+
+        device_table_->CmdBlitImage(command_buffer,
+                                    virtual_image.image,
+                                    VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL,
+                                    swapchain_image,
+                                    VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL,
+                                    1,
+                                    &blit,
+                                    VK_FILTER_NEAREST);
+
+        final_barrier_virtual.image                       = virtual_image.image;
+        final_barrier_virtual.subresourceRange.layerCount = swapchain_info->image_array_layers;
+
+        device_table_->CmdPipelineBarrier(command_buffer,
+                                          VK_PIPELINE_STAGE_TRANSFER_BIT,
+                                          VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
+                                          0,
+                                          0,
+                                          nullptr,
+                                          0,
+                                          nullptr,
+                                          1,
+                                          &final_barrier_virtual);
+
+        final_barrier_swap.image                       = swapchain_image;
+        final_barrier_swap.subresourceRange.layerCount = swapchain_info->image_array_layers;
+
+        device_table_->CmdPipelineBarrier(command_buffer,
+                                          VK_PIPELINE_STAGE_TRANSFER_BIT,
+                                          VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
+                                          0,
+                                          0,
+                                          nullptr,
+                                          0,
+                                          nullptr,
+                                          1,
+                                          &final_barrier_swap);
+
+        result = device_table_->EndCommandBuffer(command_buffer);
+        if (result != VK_SUCCESS)
+        {
+            return result;
+        }
+
+        VkPipelineStageFlags wait_stage = VK_PIPELINE_STAGE_TRANSFER_BIT;
+
+        VkSubmitInfo submit_info         = { VK_STRUCTURE_TYPE_SUBMIT_INFO };
+        submit_info.waitSemaphoreCount   = present_info->waitSemaphoreCount;
+        submit_info.pWaitSemaphores      = present_info->pWaitSemaphores;
+        submit_info.pWaitDstStageMask    = &wait_stage;
+        submit_info.commandBufferCount   = 1;
+        submit_info.pCommandBuffers      = &command_buffer;
+        submit_info.signalSemaphoreCount = 1;
+        submit_info.pSignalSemaphores    = &semaphore;
+
```

Comment:
I think there is a cleaner way to do this. Instead of overloading commands before this to change the used layouts, you could use kImageMemoryBarriers here in the vkCmdPipelineBarrier calls to change the layouts.

Reply:
Good idea! Fixed. Added a new commit `Combine two CmdPipelineBarrier`. Thanks,